src/Event/DataObjectEvent.php

```diff
@@ -0,0 +1,180 @@
+     */
+    private readonly int $timestamp;
+
+    /**
+     * @param class-string<T> $objectClass The class name of the affected DataObject
+     * @param int             $objectID    The ID of the affected DataObject
+     * @param Operation       $operation   The type of operation performed
+     * @param int|null        $version     The version number (for versioned objects)
+     * @param int|null        $memberID    The ID of the member who performed the operation
+     */
+    public function __construct(
+        private readonly string $objectClass,
+        private readonly int $objectID,
+        private readonly Operation $operation,
+        private readonly ?int $version = null,
+        private readonly ?int $memberID = null
+    ) {
+        $this->timestamp = time();
+    }
+
+    /**
+     * Get the ID of the affected DataObject
+     */
+    public function getObjectID(): int
+    {
+        return $this->objectID;
+    }
+
+    /**
+     * Get the class name of the affected DataObject
+     */
+    public function getObjectClass(): string
+    {
+        return $this->objectClass;
+    }
+
+    /**
+     * Get the type of operation performed
+     */
+    public function getOperation(): Operation
+    {
+        return $this->operation;
+    }
+
+    /**
+     * Get the version number (for versioned objects)
+     */
+    public function getVersion(): ?int
+    {
+        return $this->version;
+    }
+
+    /**
+     * Get the ID of the member who performed the operation
+     */
+    public function getMemberID(): ?int
+    {
+        return $this->memberID;
+    }
+
+    /**
+     * Get the timestamp when the event was created
+     */
+    public function getTimestamp(): int
+    {
+        return $this->timestamp;
+    }
+
+    /**
+     * Get the DataObject associated with this event
+     *
+     * @phpstan-return T|null
+     *
+     * @param bool $useVersion If true and the object is versioned, retrieves the specific version that was affected
+     *                         Note: This may return null if the object has been deleted since the event was created
+     */
+    public function getObject(bool $useVersion = false): ?DataObject
+    {
+        if (!$this->objectID) {
+            return null;
+        }
+
+        if (!$useVersion || empty($this->version)) {
+            return DataObject::get_by_id($this->objectClass, $this->objectID, false);
+        }
+
+        return Versioned::get_version($this->objectClass, $this->objectID, $this->version);
+    }
+
+    /**
+     * Get the Member who performed the operation
+     *
+     * Note: This may return null if the member has been deleted since the event was created
+     * or if the operation was performed by a system process
+     */
+    public function getMember(): ?Member
+    {
+        if (!$this->memberID) {
+            return null;
+        }
+
+        return Member::get()->byID($this->memberID);
+    }
```

Comment:
This might not play nice with our own application. Our implementation has a `User` object that is either a: `Member` or a `TrackedGuest`. I don't think our application functions without either one.